### PR TITLE
Command Line Interface

### LINKED ISSUE #451: Support Command-Line Features
For using outside Haskell (cc @nilehmann)

- [x] quiet mode       `--quiet`
- [x] read from stdin  `--stdin`
- [x] write json       `--json`
- [ ] tag

## Priorities 

1. read inputs from `stdin`
2. write outputs to `stdout` 
3. write json
4. tag

## READ Inputs from stdin

```
$ fixpoint -q -in
... // read horn format query from STDIN
```

## WRITE Output in JSON format

* `-q`  ONLY outputs SAFE/UNSAFE (n-constraints checked)

* `-json` ONLY outputs JSON  (like `-q` but with JSON output)

## JSON 

```json
{ 
  result : "safe" | "unsafe";
  errors : array<tag>;
} 
``` 

## TAG Query Format 

```smt
(qualif Eq ((x int) (y int)) (x = y))

(qualif GeZero ((v int)) (v >= 0))

(var $k ((int) (int) (int)))

(constraint 
  (forall ((l$n0 int) ((l$n0 >= 0))) 
    (and 
      (forall ((l$n1 int) (l$n0 = l$n1))
        (forall ((l$i1 int) ($k l$i1 l$n0 l$n1))
          (forall ((l$r1 int) ((l$r1 >= l$i1)))
            (forall ((l$1 bool) ((l$1 <=> (l$i1 <= l$n1))))
              (and 
                (forall ((_ int) (l$1))
                  (forall ((l$2 int) ((l$2 = (l$i1 + 1))))
                    (forall ((l$3 int) ((l$3 = (l$r1 + l$2))))
                      (and 
                        (forall ((_v int) ((_v = l$2))) (($k l$2 l$n0 l$n1 )))
                        (forall ((_v int) ((_v = l$2))) (@ (l$2 >= 0) "TAG-1"))
                        (forall ((_v int) ((_v = l$n1))) (@ (l$n0 = l$n1) "TAG-2"))
                        (forall ((_v int) ((_v = l$3))) ((@ (_v >= l$2) "TAG-3")))))))
                (forall ((_ int) ((not l$1))) 
                  (forall ((_v int) ((_v = l$r1)))
                    ((@ (_v >= l$n0) "TAG-4")))))))))
      (forall ((l$6 int) ((l$6 = 0))) 
        (forall ((l$7 int) ((l$7 = 0)))
          (and 
            (forall ((_v int) ((_v = l$7))) ((@ (_v >= l$6) "TAG-5")))
            (forall ((_v int) ((_v = l$6))) (($k l$6 l$n0 l$n0 )))
            (forall ((_v int) ((_v = l$n0))) (@ (l$n0 = l$n0) "TAG-6"))))))))
```

### PR DESCRIPTION
**Pull request by ranjitjhala**
Misc minor changes to support CLI use.

- read files from `stdin` with `--stdin`
- write result as JSON to `stdout` with `--json`

fixes #451

### CODE REVIEW

File: src/Language/Fixpoint/Solver.hs [hunk 1]
```diff
@@ -22,14 +22,14 @@ module Language.Fixpoint.Solver (
   , simplifyFInfo
 ) where
 
-import           Control.Concurrent
-import           Data.Binary
+import           Control.Concurrent                 (setNumCapabilities)
+import           Data.Binary                        (decodeFile)
+import           Data.Aeson                         (ToJSON, encode, toJSON)
```
Comment (nilehmann) [inline src/Language/Fixpoint/Solver.hs]:
```suggestion
import           Data.Aeson                         (ToJSON, encode)
import qualified Data.Text.Lazy.IO                as LT
import qualified Data.Text.Lazy.Encoding          as LT
```

File: src/Language/Fixpoint/Solver.hs [hunk 2]
```diff
@@ -58,19 +58,20 @@ solveFQ cfg = do
...
-resultExitCode :: Result SubcId -> IO ExitCode 
+resultExitCode :: (Fixpoint a, NFData a, ToJSON a) => Config -> Result a 
+               -> IO ExitCode
 ---------------------------------------------------------------------------
-resultExitCode r = do 
-  -- let str  = render $ resultDoc $!! (const () <$> stat)
-  -- putStrLn "\n"
+resultExitCode cfg r = do 
   whenNormal $ colorStrLn (colorResult stat) (statStr $!! stat)
+  when (json cfg) $ putStrLn jStr
```
Comment (nilehmann) [inline src/Language/Fixpoint/Solver.hs]:
```suggestion
  when (json cfg) $ LT.putStrLn jStr
```

File: src/Language/Fixpoint/Solver.hs [hunk 3]
```diff
@@ -58,19 +58,20 @@ solveFQ cfg = do
...
   return (eCode r)
   where 
+    jStr    = show . encode . toJSON $ r
```
Comment (nilehmann) [inline src/Language/Fixpoint/Solver.hs]:
```suggestion
    jStr    = LT.decodeUtf8 . encode $ r
```

**Comment by ranjitjhala** [comment]
aha, those little "commit suggestion" buttons would have done the trick. oh well!

**Pull request merged**